sycl/plugins/unified_runtime/ur/adapters/cuda/image.cpp

```diff
@@ -0,0 +1,1058 @@
+    array_desc.Depth = pImageDesc->depth;
+  }
+
+  ScopedContext Active(hDevice->getContext());
+
+  // Allocate a cuArray
+  if (pImageDesc->numMipLevel == 1) {
+    CUarray ImageArray;
+
+    try {
+      RetErr = UR_CHECK_ERROR(cuArray3DCreate(&ImageArray, &array_desc));
+      *phImageMem = (ur_exp_image_mem_handle_t)ImageArray;
+    } catch (ur_result_t Err) {
+      cuArrayDestroy(ImageArray);
+      return Err;
+    } catch (...) {
+      cuArrayDestroy(ImageArray);
+      return UR_RESULT_ERROR_UNKNOWN;
+    }
+  } else // Allocate a cuMipmappedArray
+  {
+    CUmipmappedArray mip_array;
+    array_desc.Flags = CUDA_ARRAY3D_SURFACE_LDST;
+
+    try {
+      RetErr = UR_CHECK_ERROR(cuMipmappedArrayCreate(&mip_array, &array_desc,
+                                                     pImageDesc->numMipLevel));
+      *phImageMem = (ur_exp_image_mem_handle_t)mip_array;
+    } catch (ur_result_t Err) {
+      cuMipmappedArrayDestroy(mip_array);
+      return Err;
+    } catch (...) {
+      cuMipmappedArrayDestroy(mip_array);
+      return UR_RESULT_ERROR_UNKNOWN;
+    }
+  }
+
+  return RetErr;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesImageFreeExp(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice,
+    ur_exp_image_mem_handle_t hImageMem) {
+  UR_ASSERT(hContext, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT((hContext->getDevice()->get() == hDevice->get()),
+            UR_RESULT_ERROR_INVALID_CONTEXT);
+  UR_ASSERT(hImageMem, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+
+  ScopedContext Active(hDevice->getContext());
+  try {
+    RetErr = UR_CHECK_ERROR(cuArrayDestroy((CUarray)hImageMem));
+  } catch (ur_result_t Err) {
+    return Err;
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+  return RetErr;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesUnsampledImageCreateExp(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice,
+    ur_exp_image_mem_handle_t hImageMem, const ur_image_format_t *pImageFormat,
+    const ur_image_desc_t *pImageDesc, ur_mem_handle_t *phMem,
+    ur_exp_image_handle_t *phImage) {
+  UR_ASSERT(hContext, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT((hContext->getDevice()->get() == hDevice->get()),
+            UR_RESULT_ERROR_INVALID_CONTEXT);
+  UR_ASSERT(hImageMem, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(pImageFormat, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pImageDesc, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(phMem, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(phImage, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+
+  unsigned int NumChannels = 0;
+  RetErr = urCalculateNumChannels(pImageFormat->channelOrder, &NumChannels);
+  if (RetErr != UR_RESULT_SUCCESS) {
+    return RetErr;
+  }
```

Comment:
Also for these cases. Isn't this pretty much just the same as using `UR_CHECK_ERROR`?